### DISCUSSION TITLE
Virtual DOM and React Fiber

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#18] [Chapter2]
## Virtual DOM

> Did you heard about `virtual DOM`? It not a real `DOM` on the browser. So user can't see virtual DOM unless before make virtual DOM to real DOM. Yeah virtual DOM is changed to real DOM, So it a kind of state before real DOM. And this is not a common thing for browser rendering process. [Here](https://github.com/honaseung/react-deep-dive/discussions/15) is a how browser render DOM. Virtual DOM is a technique of representation of DOM. Common stereotype of virtual DOM is it's absolutely faster than real DOM. It's not that fast. Just fast enough for using it. And React is using it. Then how React use it?

## The way to use virtual DOM
![스크린샷 2024-03-17 160110](https://github.com/honaseung/react-deep-dive/assets/53066347/7012d717-de2e-49ec-a620-204afb5cdee9)
> Think of magician who performance change scarf to rose. You can't see the process of chaning. Otherwise, We don't call it magic. For this magic we need scarf and rose. Scarf is DOM and rose is virtual DOM. And the magician change scarf to rose. React changes pointer (or cursor?) from DOM to virtual DOM[^doubleBuffering]. Before it Magician give a signal for audiences's attention. They try to trick them to think that signal make it happened. Like 'bibidi babidi boo'. React's signal is not exactly like this 'bibidi babidi boo'. React has a signal to know when make a new virtual DOM. When signal comes in, `React Fiber` does it's own job.

## React Fiber
![스크린샷 2024-03-27 183711](https://github.com/honaseung/react-deep-dive/assets/53066347/be131d57-599c-4460-8968-638106a2bd72)
<sup>
You can see original flow chart from [here](https://miro.com/welcomeonboard/ZndXTEFNaVJhZ2JKMW1Fc0l4eXV4ZUx1QjAzcWI4NEtuaVNzWFRGQ212QnVEUkhqR0VCaEQzRzVTNE1sMTVnaXwzNDU4NzY0NTUwNTUwMjg4Mzg4fDI=?share_link_id=51748511067).
</sup>
> React Fiber is an object which contain React Component's state. It has three elements representatively.
> 1. tag - Expression of which element of HTML.
> 2. stateNode - Fiber's reference for fiber's state (if component is changed or not? 🤔)
> 3. child, sibling, return - Expression of relation with near component (return is for parent)
>
> with these. React Fiber does three things
>
> 1. Dividing work as a unit and making a priority
> 2. Stopping work and resuming work
> 3. Doing previous work or discarding it
> * all things go on asynchronously
> when Fiber does it's own job, it starts from `beginWork()` till no child is exists. When child is no more exists, end with `completeWork()` and goes to the sibling if exists. After this, `return` to itself and let it knows job done. This job goes on till `highest parent`'s job is done.

## Commit Phase

> After Fiber's job is done. `finishWork()` is called. And then `commitWork()` is called. It's worked synchronously (From React 18, asynchronously 😀) and can't stoppable. In this phase, it works as a `double buffering` technique.

[^doubleBuffering]: paint two DOM tree and just change pointer (or cursor?) from old one to new one.

**Comment by honaseung** [comment]
가상 DOM

가상 DOM에 대해 들어보셨나요? 브라우저 상에서 실제 DOM이 아니라 가상 DOM을 실제 DOM으로 만들지 않으면 가상 DOM을 볼 수 없습니다. 네, 가상 DOM이 실제 DOM으로 됩니다. 그래서 가상 DOM은 실제 DOM이 되기 이전의 상태라고 할 수 있습니다. 그리고 이것은 브라우저 렌더링 과정에서 흔히 볼 수 있는 일이 아닙니다. 여기에 브라우저가 DOM을 렌더링하는 방법이 있습니다. 가상 DOM은 DOM을 표현하는 기술입니다. 가상 DOM의 일반적인 고정관념은 실제 DOM보다 절대적으로 빠르다는 것 입니다. 실제론 그렇게 빠르지는 않습니다. 사용하기에 충분할 정도로만 빠릅니다. 그리고 리액트가 이런 가상 DOM을 사용하고 있죠. 그럼 리액트는 어떻게 사용하나요?

가상 DOM을 사용하는 방법

마술사가 스카프를 로즈로 바꾸는 퍼포먼스를 생각해 보세요. 우리는 스카프가 로즈로 변화하는 과정을 볼 수 없어요. 그렇지 않으면 우리는 그것을 마법이라고 부르지 않아요. 이 마술을 위해서는 스카프와 로즈가 필요해요. 여기서 스카프는 DOM이고 로즈는 가상 DOM이에요. 마술사는 스카프를 로즈로 바꿀때, 리액트는 포인터(또는 커서?)를 DOM에서 가상 DOM으로 바꿉니다. 그런데 이때 바꾸기 전에 마술사는 관객의 주의를 끌기 위해 신호를 줍니다. 마술사들은 그 신호가 스카프가 로즈로 바꾸는 일을 해낸다고 생각하도록 속이려고 노력합니다. '비비디 바비디 부'라는 마법의 문장처럼. 리액트의 신호는 이 '비비디 바비디 부'와는 사뭇 다릅니다. 리액트 새로운 가상 DOM을 만들 때 알아야 하는 신호가 있어요. 신호가 들어오면 리액트 파이버는 자신의 일을 합니다.

리액트 파이버

여기에서 원본 흐름도를 볼 수 있습니다.

React Fiber는 React Component의 상태를 포함하는 객체입니다. 대표적으로 세 가지 요소가 있습니다.

tag - HTML의 요소를 표현합니다.
stateNode - Fiber 상태에 대한 Fiber의 참조(구성요소가 변경됐는지 아닌지? 🤔)
자식, 형제, 반환 - 근 성분과의 관계식(반환은 부모에 대한 것)
이것들로. 리액트 파이버는 세 가지를 합니다

단위별 업무 분담 및 우선순위 부여
작업 중지 및 작업 재개
이전 작업 수행 또는 폐기
모든 일이 비동시적으로 진행됨
Fiber가 자신의 일을 할 때, 그것은 beginWork() 호출하면서 Child가 없을 때까지 진행합니다. 아이가 더 이상 없을 때, completeWork()로 끝내고, 만약 형제가 있다면 그 형제에게 갑니다. 그 후, 다시 돌아가서 각 파이버의 일이 끝났다는 것을 알립니다. 이 작업은 가장 높은 부모까지 이어집니다.

커밋 단계

Fiber의 작업이 끝나면 finishWork()가 호출됩니다. 그런 다음 commitWork()가 호출됩니다. 동기식(React 18서 부터 비동기식 😀)으로 작동하여 중지할 수 없습니다. 이 단계에서는 이중 버퍼링 기술로 작동합니다.

**Reply by chaminju927** [reply]
마술사를 예시로 가상Dom을 설명해주시는 부분 정말 이해가 잘 되네요!! 리액으 파이버의 작업과 렌더링 까지의 프로세스를 도식화해서 올려주신 부분도 정말 큰 도움이 됐습니다 !!!
